app/models/concerns/course/search_concern.rb

```diff
@@ -0,0 +1,18 @@
+module Course::SearchConcern
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    # Search and filter courses by their titles, descriptions or user names.
+    # @param [String] keywords The keywords for filtering courses.
+    # @return [Array<Course>] The courses which match the keyword. All courses will be returned if
+    #   keyword is blank.
+    def search(keyword)
+      return all if keyword.blank?
+
+      condition = "%#{keyword}%"
+      joins { users.outer }.
+        where { (title =~ condition) | (description =~ condition) | (users.name =~ condition) }.
+        select { 'DISTINCT courses.*' }
```

Comment:
http://stackoverflow.com/questions/3408037/distinct-function-not-select-qualifier-in-postgres